src/devices/sound/nds_sound.cpp

```diff
@@ -0,0 +1,829 @@
+	save_item(STRUCT_MEMBER(m_channel, length));
+
+	save_item(STRUCT_MEMBER(m_channel, playing));
+	save_item(STRUCT_MEMBER(m_channel, adpcm_out));
+	save_item(STRUCT_MEMBER(m_channel, adpcm_index));
+	save_item(STRUCT_MEMBER(m_channel, prev_adpcm_out));
+	save_item(STRUCT_MEMBER(m_channel, prev_adpcm_index));
+	save_item(STRUCT_MEMBER(m_channel, cur_addr));
+	save_item(STRUCT_MEMBER(m_channel, cur_state));
+	save_item(STRUCT_MEMBER(m_channel, cur_bitaddr));
+	save_item(STRUCT_MEMBER(m_channel, delay));
+	save_item(STRUCT_MEMBER(m_channel, sample));
+	save_item(STRUCT_MEMBER(m_channel, lfsr));
+	save_item(STRUCT_MEMBER(m_channel, lfsr_out));
+	save_item(STRUCT_MEMBER(m_channel, counter));
+	save_item(STRUCT_MEMBER(m_channel, output));
+	save_item(STRUCT_MEMBER(m_channel, loutput));
+	save_item(STRUCT_MEMBER(m_channel, routput));
+
+	save_item(STRUCT_MEMBER(m_capture, control));
+	save_item(STRUCT_MEMBER(m_capture, dstaddr));
+	save_item(STRUCT_MEMBER(m_capture, length));
+
+	save_item(STRUCT_MEMBER(m_capture, counter));
+	save_item(STRUCT_MEMBER(m_capture, cur_addr));
+	save_item(STRUCT_MEMBER(m_capture, cur_waddr));
+	save_item(STRUCT_MEMBER(m_capture, cur_bitaddr));
+	save_item(STRUCT_MEMBER(m_capture, enable));
+
+	for (int c = 0; c < 2; c++)
+	{
+		for (int f = 0; f < 8; f++)
+			save_item(NAME(m_capture[c].fifo[f].d), (f << 1) | c);
+	}
+	save_item(STRUCT_MEMBER(m_capture, fifo_head));
+	save_item(STRUCT_MEMBER(m_capture, fifo_tail));
+	save_item(STRUCT_MEMBER(m_capture, fifo_empty));
+	save_item(STRUCT_MEMBER(m_capture, fifo_full));
+
+	save_item(NAME(m_control));
+	save_item(NAME(m_bias));
+}
+
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void nds_sound_device::device_reset()
+{
+	// TODO: registers are resetted?
+}
+
+
+//-------------------------------------------------
+//  device_clock_changed - called if the clock
+//  changes
+//-------------------------------------------------
+
+void nds_sound_device::device_clock_changed()
+{
+	m_stream->set_sample_rate(clock() / 2 / CLOCK_DIVIDE);
+}
+
+
+//-------------------------------------------------
+//  memory_space_config - return a description of
+//  any address spaces owned by this device
+//-------------------------------------------------
+
+device_memory_interface::space_config_vector nds_sound_device::memory_space_config() const
+{
+	return space_config_vector{ std::make_pair(0, &m_data_config) };
+}
+
+
+//-------------------------------------------------
+//  sound_stream_update - handle a stream update
+//-------------------------------------------------
+
+void nds_sound_device::sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs)
+{
+	outputs[0].fill(0);
+	outputs[1].fill(0);
+	for (int sampindex = 0; sampindex < outputs[0].samples(); sampindex++)
+	{
+		if (enable())
+		{
```

Comment:
Is it possible for the result of `enable()` to change per-sample?  If not, it would be better to lift it out of the loop, and only run the loop if it returns true, and zero-fill the outputs otherwise.